lib/withOnyx.d.ts

```diff
@@ -0,0 +1,36 @@
+import {IsEqual} from 'type-fest';
+import {GetOnyxValue, OnyxKey, Selector} from './types';
+
+type BaseMapping<TComponentProps, TOnyxProps> = {
+    canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+    initWithStoredValues?: boolean;
+};
+
+type Mapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends OnyxKey> = BaseMapping<TComponentProps, TOnyxProps> &
+    (
+        | (IsEqual<GetOnyxValue<TOnyxKey> | null, TOnyxProps[TOnyxProp]> extends true
+              ? {
+                    key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
+                }
+              : never)
+        | {
+              key: TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+        | {
+              key: (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+    );
+
+type OnyxPropMapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps> = {
+    [TOnyxKey in OnyxKey]: Mapping<TComponentProps, TOnyxProps, TOnyxProp, TOnyxKey>;
```

Comment:
Can we simply say that the mapping is the [ConnectOptions](https://github.com/Expensify/react-native-onyx/blob/32abd45e55c79a4b5b4666b4f07a80d93efff14d/lib/Onyx.d.ts#L17) type? We spread this mapping into the connection options we pass to Onyx.connect here. 
https://github.com/Expensify/react-native-onyx/blob/8dba043803a6269e90148519a9bc26838e7dda1e/lib/withOnyx.js#L146-L157

Maybe we need to make some updates to ConnectOptions to add extra properties, but it makes sense to me that they're the same type. Thoughts?

Reply:
@neil-marcellini I keep both types different for some reasons:

1. `withOnyx` is demanding a very complex structure in its connection options and I don't know it will work nicely in `Onyx.connect()`.
2. I understand that using `Onyx.connect()` directly from the app is different to the way it is used inside `withOnyx`. `withOnyx` automatically fills `withOnyxInstance` and some other properties that I didn't include in the typing. Looking at the app looks like only `canEvict`, `initWithStoredValues`, `key`, `selector` are used in `withOnyx`. Please let me know if this concept is wrong, but I believe that even if I keep both connection options with the same properties we won't be able to fully reuse them, due to complexity that we had to do in `withOnyx` types.

Reply:
@neil-marcellini Another question that came up right now. In API docs is saying that the `selector` option is only used with `withOnyx` -> https://github.com/Expensify/react-native-onyx/blob/main/API.md#connectmapping--number

Should we exclude it from `ConnectOptions` in `Onyx.d.ts`?

Reply:
Ok thanks for explaining your reasoning about keeping the types separate. For point 1 I would say of course it will work because we pass in into `Onyx.connect` already; but maybe I'm misunderstanding. Point 2 makes sense.

Reply:
> API docs is saying that the `selector` option is only used with `withOnyx` -> https://github.com/Expensify/react-native-onyx/blob/main/API.md#connectmapping--number
> 
> Should we exclude it from `ConnectOptions` in `Onyx.d.ts`?

Yeah sounds like that's true, all the places I see selector used are within side a block checking for withOnyxInstance.

Reply:
@neil-marcellini @blazejkustra @hayata-suenaga Will remove `selector` from connect() then.

Reply:
> For point 1 I would say of course it will work because we pass in into Onyx.connect already

@neil-marcellini these complex mapping types are complex because we're checking the selector's return value against the prop's type if there is a selector and key's corresponding onyx type against the prop type if there is no selector.